pt/lessons/specifics/ecto.md

```diff
@@ -0,0 +1,368 @@
+    select: u
+```
+
+Outros exemplos de consultas podem ser encontradas na descrição do módulo [Ecto.Query.API](http://hexdocs.pm/ecto/Ecto.Query.API.html).
+
+## Changesets
+
+Na secção anterior aprendemos como recuperar dados. Mas então como inserir e actualizá-los? Para isso precisamos de *Changesets*.
+
+Changesets cuidam da filtragem, validação, manutenção das *constraints* quando alteramos um modelo.
+
+Para este exemplo iremos nos focar no *changeset* para criação de conta de utilizador. Para começar precisamos actualizar o nosso modelo:
+
+```elixir
+defmodule ExampleApp.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Comeonin.Bcrypt, only: [hashpwsalt: 1]
+
+  schema "users" do
+    field :username, :string
+    field :encrypted_password, :string
+    field :email, :string
+    field :confirmed, :boolean, default: false
+    field :password, :string, virtual: true
+    field :password_confirmation, :string, virtual: true
+
+    timestamps
+  end
+
+  @required_fields ~w(username email password password_confirmation)
+  @optional_fields ~w()
+
+  def changeset(user, params \\ :empty) do
+    user
+    |> cast(params, @required_fields, @optional_fields)
+    |> validate_length(:password, min: 8)
+    |> validate_password_confirmation()
+    |> unique_constraint(:username, name: :email)
+    |> put_change(:encrypted_password, hashpwsalt(params[:password]))
+  end
+
+  defp validate_password_confirmation(changeset) do
+    case get_change(changeset, :password_confirmation) do
+      nil ->
+        password_mismatch_error(changeset)
+      confirmation ->
+        password = get_field(changeset, :password)
+        if confirmation == password, do: changeset, else: password_incorrect_error(changeset)
+    end
+  end
+
+  defp password_mismatch_error(changeset) do
+    add_error(changeset, :password_confirmation, "Passwords does not match")
+  end
+
+  defp password_incorrect_error(changeset) do
+    add_error(changeset, :password, "is not valid")
+  end
+end
+```
+
+Melhoramos nossa função `changeset/2` e adicionamos três novas funções auxiliares: `validate_password_confirmation/1`, `password_mismatch_error/1` e `password_incorrect_error/1`.
+
+Como o próprio nome sugere, `changeset/2` cria para nós um novo *changeset*. Nele usamos `cast/4` para converter nossos parametros para um *changeset* a partir de um conjuto de campos obrigatórios e opcionais. A seguir validamos o tamanho da palavra-passe do *changeset*, correspondência da confirmação da palavra-passe usando a nossa propria função, e a unicidade do nome de utilizador. Por último actualizamos nosso actual campo da base de dados para palavra-passe. Para tal usamos `put_change/3` para actualizar um valor no *changeset*. password confirmation match using our own function, and username uniqueness.  Finally we update our actual password database field.  For this we use `put_change/3` to update a value in the changeset.
```

Comment:
replace `Por último actualizamos nosso actual campo da base de dados para palavra-passe` for 
`Por último, atualizaremos nosso campo de password do banco de dados`.